---
title: "Logistic-RSS failure"
description: "Demonstrate that Logistic-RSS is a bad choice, contrast with Linear-RSS (i.e. SuSiE, and Logistic SuSiE (via GIBSS))"
---

**Simulation Settings**

We take 50k UKBB genotypes in an approximately 1 Mb region of chromosomse 1 `chr1_100794065_101983457`
Describe the simulation here...

```{r setup}
#project_dir <- snakemake@config$project_dir
project_dir <-  '/project2/mstephens/ktayeb/logistic-susie-snakemake'
knitr::opts_knit$set(root.dir = project_dir)
```

```{r setup2}
library(dplyr)
library(dplyr)
library(reticulate)
library(tidyr)
library(tictoc)
library(ggplot2)
library(kableExtra)

source('workflow/scripts/plotting/plot_functions.R')

# httpgd::hgd()
# NOTE: using the SER simulations right now
res_pip <- readRDS("results/ukbb_geno/chr1_100794065_101983457/ser_pip_summary.rds")
res_cs <- readRDS("results/ukbb_geno/chr1_100794065_101983457/ser_cs_summary.rds")
manifest <- readr::read_delim('config/ukbb_sim/ukbb_ser_manifest.tsv', delim='\t')
```

```{r}
# bin h2 of simulations
manifest <- manifest %>%
    mutate(h2_bin = cut(h2, breaks = c(0, 0.02, 0.1, 0.2, 0.7), labels = c("0-0.02","0.02-0.1", "0.1-0.2", "0.2-0.7")))
```

```{r rename_methods}
# Define a mapping of current method names to new human-readable names
method_mapping <- c("gibss_labf_L1eb" = "Logistic + Laplace, L=1",
                    "linear_susie_L1eb" = "Linear + Wakefield, L=1",
                    "rss_L1" = "Logistic + Wakefield (RSS), L = 1")

# Replace the values in the 'methods' column with the new names
res_pip <- res_pip %>%
    mutate(method = method_mapping[method])

res_cs <- res_cs %>%
    mutate(method = method_mapping[method])
```

```{r}
# load the SER fits and get the max z-score for each simulation

get_z_max <- function(sim_id){
    fit_path <- glue::glue('/project2/mstephens/ktayeb/logistic-susie-snakemake/results/ukbb_geno/chr1_100794065_101983457/{sim_id}/gibss_labf_L1eb.rds')
    fit <- readRDS(fit_path)
    z_max <- with(fit, max(abs(betahat / sqrt(shat2))))
    return(z_max)
}
get_z_max_safely <- purrr::safely(get_z_max, otherwise = -1)

sim_ids <- intersect(manifest$sim_id, unique(res_pip$sim_id))
max_z_tbl <- tibble(
    sim_id = sim_ids, 
    max_z = purrr::map_dbl(sim_ids, ~ get_z_max_safely(.x)$result)) %>%
    mutate(max_z_bin = cut(max_z, breaks = c(0, 10, 20, 30, Inf), labels = c("0-10","10-20", "20-30", "30+")))
```

```{r plot-funs}
threshold_var <- function(df, var, val, comp){
#   var <- as_string(ensym(var))
  filter(df, comp(.data[[var]], !!val))
}

cs_filtered_coverage <- function(res_cs, var, val, comp){
    Q_tbl <- res_cs %>%
        select(c(method, sim_id, q)) %>%
        unique() %>%
        group_by(method) %>%
        summarize(Q = sum(q))

    res_cs %>%
        left_join(Q_tbl) %>%
        threshold_var(var, val, comp) %>%
        mutate(val = val) %>%
        group_by(method) %>%
        summarize(coverage = mean(covered), n_found2 = list(table(n_found)), mean_cs_size = mean(cs_size), val = val[1], Q=Q[1], n_found = sum(n_found)) %>%
        unnest_wider(n_found2) 
}

make_plot <- function(res_cs, var, vals, op){
    f <- function(val){
        res_cs %>%
            cs_filtered_coverage(var, val, op)
    }
    filtered_coverage <- purrr::map_dfr(vals, f)

    a <- filtered_coverage %>%
        ggplot(aes(x = val, y = coverage, color = method)) +
        geom_path() + 
        geom_hline(yintercept = 0.95) +
        xlab(var)

    b <- filtered_coverage %>%
        mutate(power = n_found/Q) %>%
        mutate(power = `1`/Q) %>%
        ggplot(aes(x = val, y = power, color = method)) +
        geom_path() +
        xlab(var)

    cowplot::plot_grid(a, b)
}

```

## SER simulations

### Simulation settings

```{r}
# show unique simulation settings for SER simulations
manifest %>%
    filter(name == 'ser') %>%
    select(c(q, k, h2, rho, gamma)) %>%
    group_by_all() %>%
    summarize(n = n()) %>%
    rmarkdown::paged_table()
```


#### CS coverage

```{r}
res_cs %>%
    group_by(method, rho) %>%
    summarize(coverage = mean(covered), n_found = list(table(n_found)), mean_cs_size = mean(cs_size)) %>%
    unnest_wider(n_found) %>%
    kbl(caption = "CS Coverage by fraction of heritability in locus") %>%
    kable_classic(full_width = F, html_font = "Cambria") 
```


```{r}
res_cs %>%
    left_join(max_z_tbl) %>%
    filter(rho == 1) %>%
    group_by(method, max_z_bin) %>%
    summarize(coverage = mean(covered), n_found = list(table(n_found)), mean_cs_size = mean(cs_size)) %>%
    unnest_wider(n_found) %>%
    kbl(caption = "CS Coverage by fraction of heritability in locus") %>%
    kable_classic(full_width = F, html_font = "Cambria") 
```


### Power vs FDP (PIP ordering)

When $\rho = 1$ we see what we would hope for: the correct model performs best, (although, Linsear + Wakefield, i.e. the usual linear SER works well too), and Wakefield does worse.

```{r fig.asp=0.5}
# fdp vs power
fdp_plot <- res_pip %>%
    unnest_longer(c(pip, causal)) %>%
    ggplot(aes(pip, causal, color=method)) + 
    stat_fdp_power(geom='path', max_fdp=1.0) +
    labs(x = 'FDP', y = 'Power')

# reminder gamma is variance explained per snp (on liability scale)
fdp_plot + facet_wrap(vars(rho))
```

```{r fig.asp=0.3}
fdp_plot <- res_pip %>%
    filter(rho == 1) %>%
    left_join(max_z_tbl) %>%
    filter(!is.na(max_z)) %>%
    unnest_longer(c(pip, causal)) %>%
    ggplot(aes(pip, causal, color=method)) + 
    stat_fdp_power(geom='path', max_fdp=0.7) +
    labs(x = 'FDP', y = 'Power')

fdp_plot + facet_wrap(vars(max_z_bin), nrow=1)
```

```{r fig.asp=0.3}
fdp_plot <- res_pip %>%
    right_join(manifest) %>%
    filter(rho == 1) %>%
    left_join(max_z_tbl) %>%
    filter(!is.na(max_z)) %>%
    unnest_longer(c(pip, causal)) %>%
    ggplot(aes(pip, causal, color=method)) + 
    stat_fdp_power(geom='path', max_fdp=0.7) +
    labs(x = 'FDP', y = 'Power')

fdp_plot + facet_wrap(vars(h2_bin), nrow=1)
```

```{r fig.asp=0.3}
fdp_plot <- res_pip %>%
    right_join(manifest) %>%
    filter(rho == 1) %>%
    left_join(max_z_tbl) %>%
    filter(!is.na(max_z)) %>%
    unnest_longer(c(pip, causal)) %>%
    ggplot(aes(pip, causal, color=method)) + 
    stat_fdp_power(geom='path', max_fdp=0.7) +
    labs(x = 'FDP', y = 'Power')

fdp_plot + facet_wrap(vars(k), nrow=1)
```

As far as the *ordering* of PIPs is concerned, we need to get to very large $z$-values to start making a difference


```{r}
pip_calibration_plot <- res_pip %>%
    filter(name == 'ser') %>%
    left_join(max_z_tbl) %>%
    unnest_longer(c(pip, causal)) %>%
    filter(!is.na(max_z_bin)) %>%
    ggplot(aes(pip, causal)) +
    stat_pip_calibration(min_size = 10) +
    geom_abline(intercept = 0, slope = 1, color='red') +
    theme_bw() +
    labs(x='PIP', y='Frequency')

pip_calibration_plot + facet_grid(cols = vars(method), rows = vars(rho))
```

```{r eval=FALSE}
manifest %>%
    left_join(max_z_tbl) %>%
    group_by(rho, max_z_bin) %>%
    summarize(n = n())

res_cs %>%
    left_join(max_z_tbl) %>%
    filter(rho == 0.1, max_z_bin == '20+') %>%
    select(sim_id, idx, alpha, cs, cs_size, covered)

res_pip %>%
    left_join(max_z_tbl) %>%
    #filter(method == 'Logistic + Laplace, L=1')%>%
    filter(rho == 0.1)%>%
    select(method, sim_id, max_z_bin, pip, causal) %>%
    rowwise() %>%
    mutate(hit = which.max(pip) == which.max(causal)) %>%
    ungroup() %>%
    group_by(method) %>%
    summarize(mean(hit))

res_pip %>%
    # filter(method == 'Logistic + Laplace, L=1') %>%
    rowwise() %>%
    mutate(
        max_pip = max(pip), 
        causal_pip = pip[which(causal == 1)],
        max_pip_idx = which.max(pip),
        idx = which.max(causal),
        hit = max_pip_idx == idx
    ) %>%
    select(sim_id, method, max_pip, causal_pip, max_pip_idx, idx, hit) %>%
    filter(hit)

res_pip %>%
    filter(sim_id == 'b3b58') %>%
    mutate(
        max_pip = max(pip), 
        causal_pip = pip[which(causal == 1)],
        max_pip_idx = which.max(pip),
        idx = which.max(causal),
        hit = max_pip_idx == idx
    ) %>%
    select(sim_id, method, max_pip, causal_pip, max_pip_idx, idx, hit)

simid <- 'b3b58'
sim <- readRDS(glue::glue('results/ukbb_geno/chr1_100794065_101983457/{simid}/sim.rds'
))
fit <- readRDS(glue::glue('results/ukbb_geno/chr1_100794065_101983457/{simid}/gibss_labf_L1eb.rds'))
fit$pip <- as.vector(fit$alpha)

plot(as.vector(fit$lbf)); abline(v = sim$idx)

X <- readRDS('results/ukbb_geno/chr1_100794065_101983457/genotype.rds')

x1 <- X[, sim$idx]
x2 <- X[, which.max(fit$pip)]
summary(glm(sim$y ~x1, family='binomial'))$coef
summary(glm(sim$y ~x2, family='binomial'))$coef

y_aug <- c(sim$y, c(1, 1, 0, 0))
x1_aug <- c(x1, c(1, 0, 1, 0))
x2_aug <- c(x2, c(1, 0, 1, 0))
summary(glm(y_aug ~x1_aug, family='binomial'))$coef
summary(glm(y_aug ~x2_aug, family='binomial'))$coef
as_tibble(sim)
```


```{r}
res_pip2 <- readRDS("results/ukbb_geno/chr1_100794065_101983457/main_pip_summary.rds")
res_cs2 <- readRDS("results/ukbb_geno/chr1_100794065_101983457/main_cs_summary.rds")
manifest2 <- readr::read_delim('config/ukbb_sim/ukbb_main_manifest.tsv', delim='\t')

res_cs2 %>% 
    select(method, sim_id, cs_size, cs, covered, idx) %>% 
    filter(cs_size < 200) %>% 
    group_by(method) %>% 
    summarize(coverage = mean(covered), mean_cs_size = mean(cs_size)) %>%
    kbl(caption = "CS Coverage, q=3 causal variants") %>%
    kable_classic(full_width = F, html_font = "Cambria") 
```

```{r}
# fdp vs power
fdp_plot <- res_pip2 %>%
    unnest_longer(c(pip, causal)) %>%
    ggplot(aes(pip, causal, color=method)) + 
    stat_fdp_power(geom='path', max_fdp=1.0) +
    labs(x = 'FDP', y = 'Power')

# reminder gamma is variance explained per snp (on liability scale)
fdp_plot
```

```{r exit}
knitr::knit_exit()
```